bootstrap.sh

```diff
@@ -0,0 +1,194 @@
+    -D CMAKE_SHARED_LINKER_FLAGS=-resource-dir=$RESOURCE \
+    -D LIBUNWIND_ENABLE_STATIC=NO \
+    -D LIBUNWIND_USE_COMPILER_RT=YES \
+    -D LIBUNWIND_INCLUDE_DOCS=NO \
+    -D LLVM_ENABLE_PROJECTS="libunwind" \
+    -D LLVM_TARGETS_TO_BUILD="X86;" \
+    -G Ninja \
+    -S ..
+  ninja libunwind.so
+  DESTDIR=$SYSROOT ninja install
+  popd
+}
+
+# Saleem: c++abi is the libc equivalent for c++: it is the language runtime;
+# libc++ is the C++ standard library
+function build_libcxxabi () {
+  SYSROOT=$(readlink -f sysroot/)
+  RESOURCE=$(readlink -f sysroot/usr/local)
+  LIBCXX=$(readlink -f llvm-project/libcxx)
+  CC=$(which clang)
+  CXX=$(which clang++)
+
+  # TODO: a hack
+  pushd sysroot/usr/local/lib
+  if [[ ! -e crtbeginS.o ]]; then
+    ln -s linux/clang_rt.crtbegin-x86_64.o crtbeginS.o
+  fi
+  if [[ ! -e crtendS.o ]]; then
+    ln -s linux/clang_rt.crtend-x86_64.o crtendS.o
+  fi
+  popd
+
+  #rm -rf llvm-project/libcxxabi/build #
+  mkdir -p llvm-project/libcxxabi/build
+  pushd llvm-project/libcxxabi/build
+  # TODO: link with lld
+  cmake -D CMAKE_BUILD_TYPE=Release \
+    -D CMAKE_CXX_COMPILER_WORKS=YES \
+    -D CMAKE_CXX_COMPILER=$CXX \
+    -D CMAKE_CXX_COMPILER_TARGET=x86_64-unknown-linux-musl \
+    -D CMAKE_C_COMPILER=$CC \
+    -D CMAKE_C_COMPILER_TARGET=x86_64-unknown-linux-musl \
+    -D LIBCXXABI_ENABLE_STATIC=NO \
+    -D LIBCXXABI_LIBCXX_INCLUDES=$LIBCXX/include \
+    -D LIBCXXABI_LIBCXX_PATH=$LIBCXX \
+    -D LIBCXXABI_TARGET_TRIPLE=x86_64-unknown-linux-musl \
+    -D LIBCXXABI_USE_COMPILER_RT=YES \
+    -D LIBCXXABI_USE_LLVM_UNWINDER=YES \
+    -D LIBCXXABI_SYSROOT=$RESOURCE \
+    -D LIBCXXABI_SUPPORTS_NOSTDLIBXX_FLAG=NO \
+    -D CMAKE_EXE_LINKER_FLAGS="-rtlib=compiler-rt -resource-dir=$RESOURCE --sysroot=$RESOURCE" \
+    -G Ninja \
+    -S ..
+    #-D CMAKE_SHARED_LINKER_FLAGS=-resource-dir=$RESOURCE \
+    #-D LLVM_ENABLE_PROJECTS="libcxxabi" \
+    #-D LLVM_TARGETS_TO_BUILD="X86;" \
+    # -D CMAKE_CXX_COMPILER_WORKS=YES seems like a hack, but we can't test
+    # what we haven't built yet.
+    #--debug-trycompile \
+  ninja libc++abi.so
+  DESTDIR=$SYSROOT ninja install
+  popd
+}
+
+function build_libcxx () {
+  CC=$(which clang)
+  CXX=$(which clang++)
+  SYSROOT=$(readlink -f sysroot/)
+  RESOURCE=$(readlink -f sysroot/usr/local)
+
+  #rm -rf llvm-project/libcxx/build #
+  mkdir -p llvm-project/libcxx/build
+  pushd llvm-project/libcxx/build
+
+  # TODO: link with lld
+  cmake \
+    -D CMAKE_BUILD_TYPE=Release \
+    -D CMAKE_CXX_COMPILER=$CXX \
+    -D CMAKE_CXX_COMPILER_TARGET=x86_64-unknown-linux-musl \
+    -D CMAKE_CXX_COMPILER_WORKS=YES \
+    -D CMAKE_C_COMPILER=$CC \
+    -D CMAKE_C_COMPILER_TARGET=x86_64-unknown-linux-musl \
+    -D LIBCXX_ENABLE_STATIC=NO \
+    -D LIBCXX_CXX_ABI=libcxxabi \
+    -D LIBCXX_TARGET_TRIPLE=x86_64-unknown-linux-musl \
+    -D LIBCXX_SYSROOT=$SYSROOT/usr/local \
+    -D LIBCXX_HAS_MUSL_LIBC=YES \
+    -D LIBCXX_USE_COMPILER_RT=YES \
+    --debug-trycompile \
```

Comment:
```suggestion
    -D LIBCXX_INCLUDE_BENCHMARKS=NO \
    -D LIBCXX_SUPPORTS_NOSTDLIBXX_FLAG=NO \
    -D LIBCXX_BUILTINS_LIBRARY="${SYSROOT}/usr/local/lib/linux/libclang_rt.builtins-x86_64.a" \```

Reply:
just `-D LIBCXX_SUPPORTS_NOSTDLIBXX_FLAG=NO` allows libc++.so to build for me. Not sure about `LIBCXX_BUILTINS_LIBRARY`

Reply:
If that works, its better!